airbyte-integrations/connectors/source-convex/source_convex/source.py

```diff
@@ -0,0 +1,179 @@
+            "delta_has_more": self._delta_has_more,
+        }
+
+    @state.setter
+    def state(self, value: Mapping[str, Any]):
+        self._snapshot_cursor_value = value["snapshot_cursor"]
+        self._snapshot_has_more = value["snapshot_has_more"]
+        self._delta_cursor_value = value["delta_cursor"]
+        self._delta_has_more = value["delta_has_more"]
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        # Inner level of pagination shares the same state as outer,
+        # and returns None to indicate that we're done.
+        return self.state if self._delta_has_more else None
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        if self._snapshot_has_more:
+            return "/api/list_snapshot"
+        else:
+            return "/api/document_deltas"
+
+    def parse_response(
+        self,
+        response: requests.Response,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping]:
+        resp_json = response.json()
+        if self._snapshot_has_more:
+            self._snapshot_cursor_value = resp_json["cursor"]
+            self._snapshot_has_more = resp_json["hasMore"]
+            self._delta_cursor_value = resp_json["snapshot"]
+        else:
+            self._delta_cursor_value = resp_json["cursor"]
+            self._delta_has_more = resp_json["hasMore"]
+        return resp_json["values"]
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = {"tableName": self.table_name}
+        if self._snapshot_has_more:
+            if self._snapshot_cursor_value:
+                params["cursor"] = self._snapshot_cursor_value
+            if self._delta_cursor_value:
+                params["snapshot"] = self._delta_cursor_value
+        else:
+            if self._delta_cursor_value:
+                params["cursor"] = self._delta_cursor_value
+        return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]):
+        """
+        This is supposedly deprecated, but it's still used by AbstractSource to update state between calls to `read_records`.
+        """
+        return self.state
+
+    def read_records(self, *args, **kwargs):
+        for record in super().read_records(*args, **kwargs):
+            ts_ns = record["_ts"]
+            ts_seconds = ts_ns / 1000000000.0  # convert from nanoseconds.
+            # equivalent of java's `new Timestamp(transactionMillis).toInstant().toString()`
+            ts_datetime = datetime.fromtimestamp(ts_seconds)
```

Comment:
we should use `datetime.utcfromtimestamp` since `fromtimestamp` uses the local timezone (i think?)